Bug: `explicit_self` false positive with string interpolation

### New Issue Checklist

- [x] I've Updated SwiftLint to the latest version.
- [x] I've searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues).

### Bug Description

The [`explicit_self`](https://realm.github.io/SwiftLint/explicit_self.html) **analyzer** rule reports a violation (“Instance variables and functions should be explicitly accessed with `self.`”) even when the member is **already** accessed with `self.` inside string interpolation `\(...)` (and the same pattern with raw-string interpolation `\#(...)`).

**Expected:** No violation; `self` is explicit.

**Actual:** `explicit_self` reports a violation (false positive).

`explicit_self` only runs with **`swiftlint analyze`** (compiler invocations required), not `swiftlint lint`.

Minimal example:

```swift
class StringInterpolation {
    let foo = "foo"

    var description: String {
        return "StringInterpolation{foo: \(self.foo)}"
    }

    var rawStringLiteral: String {
        return #"StringInterpolation{foo: \#(self.foo)} using "raw" string literals."#
    }
}
```

Full repro (Swift package, `xcodebuild` log + analyze): https://github.com/jffmrk/swiftlint-explicit-self-string-interpolation

Example command output:

```text
$ swiftlint analyze --compiler-log-path xcodebuild.log
Analyzing 'StringInterpolation.swift' (1/1)
/Users/jmark/src/github/swiftlint-explicit-self-string-interpolation/Sources/ExplicitSelfStringInterpolationBug/StringInterpolation.swift:5:43: error: Explicit Self Violation: Instance variables and functions should be explicitly accessed with 'self.' (explicit_self)
/Users/jmark/src/github/swiftlint-explicit-self-string-interpolation/Sources/ExplicitSelfStringInterpolationBug/StringInterpolation.swift:9:45: error: Explicit Self Violation: Instance variables and functions should be explicitly accessed with 'self.' (explicit_self)
```

(Line/column may vary slightly by Swift/Xcode version.)

### Environment

* **SwiftLint version (run `swiftlint version` to be sure):** 0.63.2
* **Xcode version (run `xcodebuild -version` to be sure):** Xcode 26.4 (Build version 17E192)
* **Installation method used (Homebrew, CocoaPods, building from source, etc):** Homebrew (`/opt/homebrew/bin/swiftlint`)
* **Configuration file:**

```yml
analyzer_rules:
  - explicit_self

explicit_self:
  severity: error
```

**Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?** No — single `.swiftlint.yml` at the repository root.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -112,6 +112,11 @@
   existing `@TestInjected` and `@TestWeakly` exclusions.  
   [William-Laverty](https://github.com/William-Laverty)
   [#5803](https://github.com/realm/SwiftLint/pull/5803)
+  
+* Fix `explicit_self` false positives around string interpolation.  
+  [jffmrk](https://github.com/jffmrk)
+  [SimplyDanny](https://github.com/SimplyDanny)
+  [#6611](https://github.com/realm/SwiftLint/issues/6611)
 
 * Properly taint variables in tuples for `unneeded_escaping` rule.  
   [SimplyDanny](https://github.com/SimplyDanny)

--- a/Source/SwiftLintBuiltInRules/Rules/Style/ExplicitSelfRule.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/ExplicitSelfRule.swift
@@ -1,5 +1,6 @@
 import Foundation
 import SourceKittenFramework
+import SwiftSyntax
 
 struct ExplicitSelfRule: CorrectableRule, AnalyzerRule {
     var configuration = SeverityConfiguration<Self>(.warning)
@@ -38,105 +39,51 @@ struct ExplicitSelfRule: CorrectableRule, AnalyzerRule {
     }
 
     private func violationRanges(in file: SwiftLintFile, compilerArguments: [String]) -> [NSRange] {
-        guard compilerArguments.isNotEmpty else {
+        guard compilerArguments.isNotEmpty, let filePath = file.path else {
             Issue.missingCompilerArguments(path: file.path, ruleID: Self.identifier).print()
             return []
         }
+        let contents = file.stringView
+        return DeclReferenceOffsetCollector(filePath: filePath, compilerArguments: compilerArguments)
+            .walk(tree: file.syntaxTree, handler: \.offsets)
+            .compactMap { contents.byteRangeToNSRange(ByteRange(location: $0, length: 0)) }
+    }
+}
 
-        let allCursorInfo: [[String: any SourceKitRepresentable]]
+private final class DeclReferenceOffsetCollector: SyntaxVisitor {
+    private(set) var offsets: [ByteCount] = []
+
+    private let filePath: String
+    private let compilerArguments: [String]
+
+    private static let kindsToFind: Set = [
+        "source.lang.swift.ref.function.method.instance",
+        "source.lang.swift.ref.var.instance",
+    ]
+
+    init(filePath: String, compilerArguments: [String]) {
+        self.filePath = filePath
+        self.compilerArguments = compilerArguments
+        super.init(viewMode: .sourceAccurate)
+    }
+
+    override func visitPost(_ node: DeclReferenceExprSyntax) {
+        guard node.keyPathInParent != \MemberAccessExprSyntax.declName,
+              node.keyPathInParent != \KeyPathPropertyComponentSyntax.declName else {
+            return
+        }
+        let cursorInfoRequest = Request.cursorInfoWithoutSymbolGraph(
+            file: filePath,
+            offset: ByteCount(node.baseName.positionAfterSkippingLeadingTrivia),
+            arguments: compilerArguments
+        )
         do {
-            let byteOffsets = try binaryOffsets(file: file, compilerArguments: compilerArguments)
-            allCursorInfo = try file.allCursorInfo(compilerArguments: compilerArguments,
-                                                   atByteOffsets: byteOffsets)
+            let cursorInfo = try cursorInfoRequest.sendIfNotDisabled()
+            if let kind = SourceKittenDictionary(cursorInfo).kind, Self.kindsToFind.contains(kind) {
+                offsets.append(ByteCount(node.baseName.positionAfterSkippingLeadingTrivia))
+            }
         } catch {
             queuedPrintError(String(describing: error))
-            return []
-        }
-
-        let cursorsMissingExplicitSelf = allCursorInfo.filter { cursorInfo in
-            guard let kindString = cursorInfo["key.kind"] as? String else { return false }
-            return kindsToFind.contains(kindString)
-        }
-
-        guard cursorsMissingExplicitSelf.isNotEmpty else {
-            return []
-        }
-
-        let contents = file.stringView
-
-        return cursorsMissingExplicitSelf.compactMap { cursorInfo in
-            guard let byteOffset = (cursorInfo["swiftlint.offset"] as? Int64).flatMap(ByteCount.init) else {
-                Issue.genericWarning("Cannot convert offsets in '\(Self.identifier)' rule.").print()
-                return nil
-            }
-
-            return contents.byteRangeToNSRange(ByteRange(location: byteOffset, length: 0))
         }
     }
-}
-
-private let kindsToFind: Set = [
-    "source.lang.swift.ref.function.method.instance",
-    "source.lang.swift.ref.var.instance",
-]
-
-private extension SwiftLintFile {
-    func allCursorInfo(compilerArguments: [String], atByteOffsets byteOffsets: [ByteCount]) throws
-        -> [[String: any SourceKitRepresentable]] {
-        try byteOffsets.compactMap { offset in
-            if isExplicitAccess(at: offset) { return nil }
-            let cursorInfoRequest = Request.cursorInfoWithoutSymbolGraph(
-                file: self.path!, offset: offset, arguments: compilerArguments
-            )
-            var cursorInfo = try cursorInfoRequest.sendIfNotDisabled()
-
-            // Accessing a `projectedValue` of a property wrapper (e.g. `self.$foo`) or the property wrapper itself
-            // (e.g. `self._foo`) results in an incorrect `key.length` (it does not account for the identifier
-            // prefixes `$` and `_`), while `key.name` contains the prefix. Hence we need to check for explicit access
-            // at a corrected offset as well.
-            var prefixLength: Int64 = 0
-            let sourceKittenDictionary = SourceKittenDictionary(cursorInfo)
-            if sourceKittenDictionary.kind == "source.lang.swift.ref.var.instance",
-               let name = sourceKittenDictionary.name,
-               let length = sourceKittenDictionary.length {
-                prefixLength = Int64(name.count - length.value)
-                if prefixLength > 0, isExplicitAccess(at: offset - ByteCount(prefixLength)) {
-                    return nil
-                }
-            }
-
-            cursorInfo["swiftlint.offset"] = Int64(offset.value) - prefixLength
-            return cursorInfo
-        }
-    }
-
-    private func isExplicitAccess(at location: ByteCount) -> Bool {
-        stringView.substringWithByteRange(ByteRange(location: location - 1, length: 1))! == "."
-    }
-}
-
-private extension StringView {
-    func recursiveByteOffsets(_ dict: [String: Any]) -> [ByteCount] {
-        let cur: [ByteCount]
-        if let line = dict["key.line"] as? Int64,
-           let column = dict["key.column"] as? Int64,
-           let kindString = dict["key.kind"] as? String,
-           kindsToFind.contains(kindString),
-           let offset = byteOffset(forLine: line, bytePosition: column) {
-            cur = [offset]
-        } else {
-            cur = []
-        }
-        if let entities = dict["key.entities"] as? [[String: Any]] {
-            return entities.flatMap(recursiveByteOffsets) + cur
-        }
-        return cur
-    }
-}
-
-private func binaryOffsets(file: SwiftLintFile, compilerArguments: [String]) throws -> [ByteCount] {
-    let absoluteFile = file.path!.bridge().absolutePathRepresentation()
-    let index = try Request.index(file: absoluteFile, arguments: compilerArguments).sendIfNotDisabled()
-    let binaryOffsets = file.stringView.recursiveByteOffsets(index)
-    return binaryOffsets.sorted()
 }

--- a/Source/SwiftLintBuiltInRules/Rules/Style/ExplicitSelfRuleExamples.swift
+++ b/Source/SwiftLintBuiltInRules/Rules/Style/ExplicitSelfRuleExamples.swift
@@ -35,6 +35,43 @@ struct ExplicitSelfRuleExamples {
             A(p1: 10).$p1
         }
         """),
+        Example("""
+        class StringInterpolation {
+            let foo = "foo"
+
+            var description: String {
+                return "StringInterpolation{foo: \\(self.foo)}"
+            }
+        }
+        """, testWrappingInString: false),
+        Example("""
+        class StringInterpolationRawStringLiteral {
+            let foo = "foo"
+
+            var description: String {
+                return #"StringInterpolation{foo: \\#(self.foo)}"#
+            }
+        }
+        """, testWrappingInString: false),
+        Example("""
+        class LocalStringInterpolation {
+            var bar: String
+
+            init() {
+                let a = "a"
+                let b = "b"
+                self.bar = "\\(a)\\(b)".uppercased()
+            }
+        }
+        """, testWrappingInString: false),
+        Example("""
+        class StringConcatenation {
+            var description: String {
+                let number = 1
+                return "\\(number)" + " count"
+            }
+        }
+        """, testWrappingInString: false),
     ]
 
     static let triggeringExamples = [
@@ -81,22 +118,42 @@ struct ExplicitSelfRuleExamples {
             A(p1: 10).$p1
         }
         """),
+        Example("""
+        class StringInterpolation {
+            let foo = "foo"
+
+            var description: String {
+                return "StringInterpolation{foo: \\(↓foo)}"
+            }
+        }
+        """, testWrappingInString: false),
+        Example("""
+        class StringInterpolationRawStringLiteral {
+            let foo = "foo"
+
+            var description: String {
+                return #"StringInterpolation{foo: \\#(↓foo)}"#
+            }
+        }
+        """, testWrappingInString: false),
     ]
 
     static let corrections = [
         Example("""
         struct A {
-            func f1() {}
-            func f2() {
-                ↓f1()
+            func f1() -> Int { 1 }
+            func f2() -> Int { 2 }
+            func f3() -> Int {
+                ↓f1() + ↓f2()
             }
         }
         """):
         Example("""
         struct A {
-            func f1() {}
-            func f2() {
-                self.f1()
+            func f1() -> Int { 1 }
+            func f2() -> Int { 2 }
+            func f3() -> Int {
+                self.f1() + self.f2()
             }
         }
         """),
@@ -169,5 +226,39 @@ struct ExplicitSelfRuleExamples {
             A(p1: 10).$p1
         }
         """),
+        Example("""
+        class StringInterpolation {
+            let foo = "foo"
+
+            var description: String {
+                return "StringInterpolation{foo: \\(↓foo)}"
+            }
+        }
+        """, testWrappingInString: false): Example("""
+        class StringInterpolation {
+            let foo = "foo"
+
+            var description: String {
+                return "StringInterpolation{foo: \\(self.foo)}"
+            }
+        }
+        """, testWrappingInString: false),
+        Example("""
+        class StringInterpolationRawStringLiteral {
+            let foo = "foo"
+
+            var description: String {
+                return #"StringInterpolation{foo: \\#(↓foo)}"#
+            }
+        }
+        """, testWrappingInString: false): Example("""
+        class StringInterpolationRawStringLiteral {
+            let foo = "foo"
+
+            var description: String {
+                return #"StringInterpolation{foo: \\#(self.foo)}"#
+            }
+        }
+        """, testWrappingInString: false),
     ]
 }

